README.md
# shapewear [![Build Status](https://secure.travis-ci.org/elementar/shapewear.png)](http://travis-ci.org/elementar/shapewear)

Shapewear make your fat service look skinny.

Most of the inspiration for this gem came from [Savon](http://savonrb.com/) and [HTTParty](http://httparty.rubyforge.org/), 
thanks Daniel Harrington (a.k.a. rubiii) and John Nunemaker!

## Work in Progress

This gem is still in early development. It's only working for some very basic cases. Any contribution and feedback is welcome.

### Roadmap

* Add more tests;
* Add support for SOAP 1.2;
* Add cleaner integration on Rails and Sinatra (Rack middleware?).

## Installation

Shapewear is available through [Rubygems](http://rubygems.org/gems/shapewear) and can be installed via:

```
$ gem install shapewear
```

Introduction
------------

First, describe your SOAP service:

``` ruby
require "shapewear"

class MyFirstService
  include Shapewear

  wsdl_namespace 'http://services.example.com/v1'

  endpoint_url 'http://localhost:3000/my_first_service'

  operation :hello, :parameters => [[:name, String]], :returns => String
  def hello(name)
    "hello, #{name}"
  end

  operation :sum, :parameters => [[:x, Fixnum], [:y, Fixnum]], :returns => Fixnum
  def sum(x, y)
    x + y
  end

  operation :get_user_info, :parameters => [[:email, String]], :returns => { :name => String, :birthday => DateTime }
  def get_user_info(email)
    User.find_by_email(email)
  end
end
```

Then bind to your web application in a non-intrusive way.

Rails example:

``` ruby
# don't forget to write the appropriate routes
class MyFirstServiceController < ApplicationController
  def wsdl
    render :xml => MyHelloService.to_wsdl
  end

  def serve
    render :xml => MyHelloService.serve(request)
  end
end
```

Sinatra example:

``` ruby
class MySinatraApp < Sinatra::App
  get "my_first_service/wsdl" do
    content_type "application/xml"
    MyHelloService.to_wsdl
  end

  post "my_first_service" do
    content_type "application/xml"
    MyHelloService.serve(request)
  end
end
```


CHANGELOG.md
# Changelog

## 0.1.8
* Better key lookup on hashes.
* Changed #camelize to work the same way as in Rails.

## 0.1.5
* #camelize in parameters and complex types is only called if they are declared as symbols.
  This allows usage of custom character casing when writing the SOAP services.
* Parameters are now camelized with first letter as lowercase.

## 0.1.4
* Parameters are now read in the correct order, as specified in the metadata;
* Complex return types are now working correctly;
* DRYed some code;
* Added more logs.

## 0.1.3
* More fixes for .NET interoperability.

## 0.1.2
* Fixed a bug when returning SOAP:Faults.

## 0.1.1
* Added support for SOAP 1.2 Faults;
* Refactored the request handling into its own class.

## 0.1.0
* Added specs and fixed basic support for SOAP:Fault.

## 0.0.6
* Added basic support for SOAP:Fault.

## 0.0.5
* Removed the dual-namespace option. Now there's only one.
* Generating WSDL for SOAP 1.2;
* Added an option on the DSL to change the service name;
* Changed some conventions on the generated WSDL so Visual Studio .NET can understand it better.

## 0.0.4
* Fixed the samples;
* Fixed some bugs that prevented the generated WSDL to be used with Visual Studio.

## 0.0.3
* Added support for WSDL documentation;
* Added wasabi and WebMock to the specs;
* Added support for complex types on parameters, in addition to return values;
* Request handling for very simple cases.

## 0.0.2
* WSDL generation code extracted to its own module;
* Added support for complex return types;
* Fixed a bug on the String#camelize extension;
* DateTime support on parameters and return values.

## 0.0.1
* Initial, experimental version


spec/spec_helper.rb
# encoding: UTF-8

require "bundler"
Bundler.require :default, :development

require 'webmock/rspec'

RSpec.configure do |config|
end

Dir[File.join File.dirname(__FILE__), 'shapewear', 'service_definitions', '*.rb'].each { |f| require f }


spec/shapewear/dsl_spec.rb
# encoding: UTF-8

require "spec_helper"

describe Shapewear::DSL do

  let(:xmlns) { {'wsdl' => 'http://schemas.xmlsoap.org/wsdl/', 'soap' => 'http://schemas.xmlsoap.org/wsdl/soap/'} }

  describe "basic DSL" do
    it "should describe a minimal working service" do
      wsdl = MinimalWorkingService.to_wsdl
      puts wsdl

      # wsdl should be valid XML (duh!)
      expect { Nokogiri::XML(wsdl) { |c| c.strict } }.not_to raise_error

      # wasabi should be able to parse it
      wdoc = nil
      expect { wdoc = Wasabi.document wsdl }.not_to raise_error

      wdoc.namespace.should match /MinimalWorkingService/
      wdoc.soap_actions.should == [:hello_world]

      wdoc.operations[:hello_world].should_not be_nil
      wdoc.operations[:hello_world][:input].should == 'HelloWorld'
      wdoc.operations[:hello_world][:input].should match /HelloWorld$/
    end
  end

  describe "complex types DSL" do
    it "should accept complex types as input"
    it "should accept complex types as output"
    it "should accept arrays as input"
    it "should accept arrays as output"
    it "should allow definition of complex types using class introspection"
    it "should allow definition of complex types using a DSL"
  end

  describe "WSDL customization" do
    it "should allow customization of target namespace"
    it "should allow customization of schema namespace"
  end

  describe "existing WSDL" do
    it "should accept an existing WSDL"
  end
end


spec/shapewear/savon_usage_spec.rb
# encoding: UTF-8

require "spec_helper"

describe Shapewear do
  describe "usage with SOAP clients" do
    before do
      stub_request(:get, "http://services.example.com/complete/soap/wsdl") \
        .to_return :body => CompleteService.to_wsdl, :headers => { 'Content-Type' => 'application/xml' }

      stub_request(:post, "http://services.example.com/complete/soap") \
        .to_return :body => lambda { |r| CompleteService.serve(r) }, :headers => { 'Content-Type' => 'application/xml' }
    end

    describe "Savon" do
      it "should work for simple requests" do
        client = Savon::Client.new 'http://services.example.com/complete/soap/wsdl'
        response = client.request :echo_in_uppercase, :xmlns => 'http://services.example.com/v1' do
          soap.body = { :text => 'uppercase text' }
        end

        puts response.inspect
        puts response.body.inspect

        response.body[:echo_in_uppercase_response][:echo_in_uppercase_result].should == 'UPPERCASE TEXT'
      end

      it "should work for structured responses from objects" do
        client = Savon::Client.new 'http://services.example.com/complete/soap/wsdl'
        response = client.request :get_structured_data, :xmlns => 'http://services.example.com/v1' do
          soap.body = { 'ID' => 0 }
        end

        r = response.body[:get_structured_data_response][:get_structured_data_result]
        r.should be_a Hash
        r.should include :text => 'text from the struct', :xyz => 'xyz'
      end

      it "should work for structured responses from hashes" do
        client = Savon::Client.new 'http://services.example.com/complete/soap/wsdl'
        response = client.request :get_structured_data, :xmlns => 'http://services.example.com/v1' do
          soap.body = { 'ID' => 1 }
        end

        r = response.body[:get_structured_data_response][:get_structured_data_result]
        r.should be_a Hash
        r.should include :text => 'text from a hash', :xyz => 'xyz'
      end

      it "should raise SOAP 1.1 Faults" do
        client = Savon::Client.new 'http://services.example.com/complete/soap/wsdl'

        expect {
          client.request :get_structured_data, :xmlns => 'http://services.example.com/v1' do
            soap.body = { 'ID' => 55 }
          end
        }.to raise_error Savon::SOAP::Fault, "(e:Server.RuntimeError) ID must be 0 or 1"
      end

      it "should raise SOAP 1.2 Faults" do
        client = Savon::Client.new 'http://services.example.com/complete/soap/wsdl'

        expect {
          client.request :get_structured_data, :xmlns => 'http://services.example.com/v1' do
            soap.version = 2
            soap.body = { 'ID' => 55 }
          end
        }.to raise_error Savon::SOAP::Fault, "(e:Server.RuntimeError) ID must be 0 or 1"
      end
    end
  end
end

lib/shapewear.rb
# encoding: UTF-8

require 'nokogiri'

require 'shapewear/version'
require 'shapewear/logging'
require 'shapewear/dsl'
require 'shapewear/wsdl'
require 'shapewear/request'

module Shapewear
  def self.included(receiver)
    receiver.extend(Shapewear::Logging)
    receiver.extend(Shapewear::DSL)
    receiver.extend(Shapewear::WSDL)
    receiver.extend(Shapewear::Request)

    class << receiver
      def method_added(m)
        # automatically creates an operation for each method added
        operation m
      end
    end
  end
end

# defines String.camelize and String.underscore, if it is not defined by, e.g. Rails
class String
  unless ''.respond_to? :camelize
    def camelize(first_letter = :upper)
      if first_letter == :upper
        self.gsub(/\/(.?)/) { "::#{$1.upcase}" }.gsub(/(?:^|_)(.)/) { $1.upcase }
      else
        self[0].chr.downcase + self.camelize[1..-1]
      end
    end
  end

  unless ''.respond_to? :underscore
    def underscore
      word = self.dup
      word.gsub!(/::/, '/')
      word.gsub!(/([A-Z]+)([A-Z][a-z])/, '\1_\2')
      word.gsub!(/([a-z\d])([A-Z])/, '\1_\2')
      word.tr!("-", "_")
      word.downcase!
      word
    end
  end
end

class Object
  def camelize_if_symbol(first_letter = :upper)
    if self.is_a?(Symbol)
      self.to_s.camelize(first_letter)
    else
      self.to_s
    end
  end
end


lib/shapewear/dsl.rb
# encoding: UTF-8

require 'builder'

module Shapewear::DSL
  private

  def options
    @options ||= { :service_name => self.name }
  end

  def operations
    options[:operations] ||= {}
  end

  def namespaces
    options[:namespaces] ||=
      Hash.new { |_, k| raise "Namespace not defined: #{k}" } \
        .merge! 'tns' => "http://services.example.com/#{self.name}",
                'wsdl' => 'http://schemas.xmlsoap.org/wsdl/',
                'soap' => 'http://schemas.xmlsoap.org/wsdl/soap/',
                'soap12' => 'http://schemas.xmlsoap.org/wsdl/soap12/',
                'xsd' => 'http://www.w3.org/2001/XMLSchema',
                'xsi' => 'http://www.w3.org/2001/XMLSchema-instance',
                'env' => 'http://schemas.xmlsoap.org/soap/envelope/',
                'env12' => 'http://www.w3.org/2001/12/soap-envelope'
  end

  protected

  def service_name(sn)
    options[:service_name] = sn
  end

  def wsdl_namespace(ns)
    namespaces['tns'] = ns
  end

  def endpoint_url(url)
    options[:endpoint_url] = url
  end

  def operation(name, ops = {})
    h = (operations[name] ||= {}).merge! ops
    h[:public_name] ||= name.to_s.camelize
  end
end


lib/shapewear/request.rb
# encoding: UTF-8

require 'forwardable'

module Shapewear::Request
  # @param request [Rack::Request, Hash]
  def serve(request)
    RequestHandler.new(self, request).serve
  end

  class RequestHandler
    extend Forwardable

    attr_reader :soap_version, :op_node, :clazz
    def_delegators :@clazz, :namespaces, :operations, :logger

    def initialize(clazz, request)
      @clazz = clazz

      body ||= request.body if request.respond_to? :body
      body ||= request[:body] if request.is_a?(Hash)
      body ||= request.to_s

      raise "Request body could not be found" if body.nil?

      doc = Nokogiri::XML(body) { |c| c.strict } rescue raise("Request body is not a valid XML")

      # detect the SOAP version from the envelope, and find the operation element, or raise if not found
      if doc.at('/env:Envelope', namespaces)
        @soap_version = :soap11
        @op_node = doc.at("/env:Envelope/env:Body/tns:*", namespaces)
      elsif doc.at('/env12:Envelope', namespaces)
        @soap_version = :soap12
        @op_node = doc.at("/env12:Envelope/env12:Body/tns:*", namespaces)
      else
        raise "Request is not a SOAP 1.1 nor SOAP 1.2 Envelope: #{body}"
      end
    end

    def serve
      call_soap_operation
    end

    private

    def call_soap_operation
      raise "Operation node not found" if op_node.nil?

      operations.each do |k, v|
        if v[:public_name] == op_node.name
          logger.debug "Extracting parameters from operation node..."
          params = extract_parameters(v, op_node)
          logger.debug "Creating new instance of #{clazz}..."
          obj = clazz.new
          logger.debug "Calling #{k} with args: #{params.map(&:inspect) * ', '}"
          begin
            r = obj.send(k, *params)
            logger.debug "Result: #{r.inspect}"
            return serialize_soap_result v, r
          rescue => e
            logger.debug "Exception: #{e.inspect}"
            return serialize_soap_fault e
          end
        end
      end

      raise "Operation not found: #{@op_node.name}"
    end

    # Extracts all parameters from the operation node, and return as an array.
    #
    # @param op_options [Hash] The operation options.
    # @param node [Nokogiri::XML] The operation node.
    # @return [Array] The parsed parameters.
    def extract_parameters(op_options, node)
      logger.debug "Operation node: #{node.inspect}"
      r = []
      op_options[:parameters].each do |p|
        logger.debug "  Looking for: tns:#{p.first.camelize_if_symbol(:lower)}"
        v = node.xpath("tns:#{p.first.camelize_if_symbol(:lower)}", namespaces).first
        if v.nil?
          # does nothing
        elsif p.last == Fixnum
          v = v.text.to_i
        elsif p.last == DateTime
          v = DateTime.parse(v.text) # TODO: add tests
        else
          v = v.text
        end
        logger.debug "    Found: #{v.inspect}"
        r << v
      end
      r
    end

    # Serializes the result of an operation as a SOAP Envelope.
    #
    # @param op_options [Hash] The operation options.
    # @param r [Hash,Object] The operation result.
    #noinspection RubyArgCount
    def serialize_soap_result(op_options, r)
      xb = Builder::XmlMarkup.new
      xb.instruct!

      xb.Envelope :xmlns => soap_env_ns, 'xmlns:xsi' => namespaces['xsi'] do |xenv|
        xenv.Body do |xbody|
          xbody.tag! "#{op_options[:public_name]}Response", :xmlns => namespaces['tns'] do |xresp|
            if r.nil?
              xresp.tag! "#{op_options[:public_name]}Result", 'xsi:nil' => 'true'
            else
              ret = op_options[:returns] rescue nil
              case ret
                when NilClass, Class
                  xresp.tag! "#{op_options[:public_name]}Result", r
                when Hash
                  xresp.tag! "#{op_options[:public_name]}Result" do |xres|
                    ret.each do |k, v|
                      extract_and_serialize_value(xres, r, k, v)
                    end
                  end
                else
                  raise "Unsupported return type: #{ret.inspect}"
              end
            end
          end
        end
      end
    end

    # Extracts a field from an object, casts it to the appropriate type, and serializes as XML.
    #
    # @param builder [Builder::XmlMarkup] The XML builder.
    # @param obj [Hash,Object] The resulting object.
    # @param field [Symbol,String] The field to extract.
    # @param type [Class] The type to convert.
    def extract_and_serialize_value(builder, obj, field, type)
      v = if obj.is_a?(Hash)
        obj[field] or obj[field.to_sym] or obj[field.to_s.underscore] or obj[field.to_s.underscore.to_sym]
      elsif obj.respond_to?(field)
        obj.send(field)
      elsif obj.respond_to?(field.underscore)
        obj.send(field.underscore)
      else
        raise "Could not extract #{field.inspect} from object: #{obj.inspect}"
      end

      if v.nil?
        builder.tag! field.camelize_if_symbol, 'xsi:nil' => 'true'
      else
        builder.tag! field.camelize_if_symbol, v
      end
    end

    # Serializes an exception as a SOAP Envelope containing a SOAP Fault.
    #
    # @param ex [Exception] The Exception to serialize.
    # @return [String] The SOAP Envelope containing the Fault.
    #noinspection RubyArgCount
    def serialize_soap_fault(ex)
      logger.debug "Serializing SOAP Fault: #{ex.inspect}"

      xb = Builder::XmlMarkup.new
      xb.instruct!

      xb.tag! 'e:Envelope', 'xmlns:e' => soap_env_ns do |xenv|
        xenv.tag! 'e:Body' do |xbody|
          xbody.tag! 'e:Fault' do |xf|
            case soap_version
              when :soap11
                xf.faultcode "e:Server.#{ex.class.name}"
                xf.faultstring ex.message
              when :soap12
                xf.tag! 'e:Code' do |xcode|
                  xcode.tag! 'e:Value', 'e:Receiver'
                  xcode.tag! 'e:Subcode' do |xsubcode|
                    xsubcode.tag! 'e:Value', ex.class.name
                  end
                end
                xf.tag! 'e:Reason', ex.message
              else
                raise "Unsupported SOAP version: #{soap_version}"
            end
          end
        end
      end
    end

    # Returns the correct SOAP Envelope namespace.
    #
    # @return [String] The correct SOAP Envelope namespace.
    def soap_env_ns
      case soap_version
        when :soap11
          namespaces['env']
        when :soap12
          namespaces['env12']
        else
          raise "Unrecognized SOAP version: #{soap_version}"
      end
    end
  end
end


lib/shapewear/wsdl.rb
# encoding: UTF-8

require 'builder'

#noinspection RubyArgCount,RubyResolve,RubyLiteralArrayInspection
module Shapewear::WSDL
  # reference: http://www.w3.org/TR/wsdl
  def to_wsdl
    xm = Builder::XmlMarkup.new

    xm.instruct!
    xm.definitions :name => self.name, 'targetNamespace' => namespaces['tns'],
                   'xmlns' => namespaces['wsdl'],
                   'xmlns:soap' => namespaces['soap'],
                   'xmlns:soap12' => namespaces['soap12'],
                   'xmlns:xsd' => namespaces['xsd'],
                   'xmlns:tns' => namespaces['tns'] do |xdef|

      xdef.types do |xtypes|
        xtypes.schema 'xmlns' => namespaces['xsd'], :elementFormDefault => 'qualified', :targetNamespace => namespaces['tns'] do |xschema|

          # define elements for each defined method
          operations.each do |m, op_opts|
            build_type_elements_for_method(m, op_opts, xschema)
          end
        end
      end

      operations.each do |_, op_opts|
        xdef.message :name => "#{op_opts[:public_name]}SoapIn" do |xmsg|
          xmsg.part :name => :parameters, :element => "tns:#{op_opts[:public_name]}"
        end
        xdef.message :name => "#{op_opts[:public_name]}SoapOut" do |xmsg|
          xmsg.part :name => :parameters, :element => "tns:#{op_opts[:public_name]}Response"
        end
      end

      xdef.portType :name => "#{options[:service_name]}Soap" do |xpt|
        operations.each do |_, op_opts|
          xpt.operation :name => op_opts[:public_name] do |xop|
            xop.input :message => "tns:#{op_opts[:public_name]}SoapIn"
            xop.output :message => "tns:#{op_opts[:public_name]}SoapOut"
          end
        end
      end

      # bindings for soap 1.1 and 1.2
      ['', '12'].each do |sv|
        xdef.binding :name => "#{options[:service_name]}Soap#{sv}", :type => "tns:#{options[:service_name]}Soap" do |xbind|
          xbind.tag! "soap#{sv}:binding", :transport => 'http://schemas.xmlsoap.org/soap/http'
          operations.each do |_, op_opts|
            xbind.operation :name => op_opts[:public_name] do |xop|
              doc = op_opts[:documentation] rescue nil
              xop.documentation doc unless doc.nil?
              xop.tag! "soap#{sv}:operation", :soapAction => "#{namespaces['tns']}/#{op_opts[:public_name]}", :style => 'document'
              xop.input { |xin| xin.tag! "soap#{sv}:body", :use => 'literal' }
              xop.output { |xout| xout.tag! "soap#{sv}:body", :use => 'literal' }
            end
          end
        end
      end

      # service definition for soap 1.1 and 1.2
      xdef.service :name => "#{options[:service_name]}" do |xsrv|
        xsrv.documentation "WSDL auto-generated by shapewear."
        ['', '12'].each do |sv|
          xsrv.port :name => "#{options[:service_name]}Soap#{sv}", :binding => "tns:#{options[:service_name]}Soap#{sv}" do |xport|
            xport.tag! "soap#{sv}:address", :location => options[:endpoint_url]
          end
        end
      end
    end
  end

  def build_type_elements_for_method(m, op_options, xschema)
    # element for method arguments
    um = instance_method(m)

    xschema.element :name => "#{op_options[:public_name]}" do |xreq|
      xreq.complexType do |xct|
        xct.sequence do |xseq|
          params = op_options[:parameters] rescue nil
          if params.nil?
            if um.respond_to?(:parameters)
              # with Ruby 1.9, we can create the parameters with the correct names
              params = um.parameters.select { |p| p.first == :in }.map { |p| [p.first, Object] }
            else
              params = (0..um.arity).map { |i| ["arg#{i}", Object] }
            end
          end

          params.each do |p|
            t = p.last
            param_name = p.first
            param_name = param_name.camelize_if_symbol(:lower)
            if t.nil?
              xseq.element :name => param_name, :minOccurs => 0, :maxOccurs => 1, :type => 'xsd:any'
            elsif t.is_a?(Class)
              xseq.element :name => param_name, :minOccurs => 0, :maxOccurs => 1, :type => to_xsd_type(t)
            elsif t.is_a?(Hash)
              xseq.complexType do |xct2|
                xct2.sequence do |xseq2|
                  t.each do |name, type|
                    xseq2.element :name => name, :minOccurs => 0, :maxOccurs => 1, :type => to_xsd_type(type)
                  end
                end
              end
            else
              raise "Could not interpret #{t.inspect} as a return type definition"
            end
          end
        end
      end
    end

    # element for method result
    ret = op_options[:returns] rescue nil

    xschema.element :name => "#{op_options[:public_name]}Response" do |xreq|
      xreq.complexType do |xct|
        xct.sequence do |xseq|
          xseq.element :name => "#{op_options[:public_name]}Result", :minOccurs => 0, :maxOccurs => 1,
                       :type => case ret
                         when NilClass then 'xsd:any'
                         when Class then to_xsd_type(ret)
                         when Hash then "tns:#{op_options[:public_name]}Struct"
                         else
                           raise "Could not interpret #{ret.inspect} as a return type definition"
                       end
        end
      end
    end

    if ret.is_a?(Hash)
      xschema.complexType :name => "#{op_options[:public_name]}Struct" do |xctr|
        xctr.sequence do |xseqr|
          ret.each do |name, type|
            name = name.camelize_if_symbol
            xseqr.element :name => name, :minOccurs => 0, :maxOccurs => 1, :type => to_xsd_type(type)
          end
        end
      end
    end
  end

  # @param t [Class]
  def to_xsd_type(t)
    return 'xsd:string' if t == String
    return 'xsd:integer' if t == Fixnum
    return 'xsd:dateTime' if t == DateTime
    return 'xsd:any' if t.nil? || t == Object
    raise "Could not convert type #{t} to a valid XSD type"
  end
end
